contracts/walrus/sources/blob_manager.move

```diff
@@ -0,0 +1,833 @@
+): u32 {
+    // If no extension possible after capping, return 0.
+    if (effective_extension == 0) {
+        return 0
+    };
+
+    // Get available WAL balance from stash.
+    let available_wal = self.coin_stash.wal_balance();
+    if (available_wal == 0) {
+        return 0
+    };
+
+    // Get current storage info.
+    let capacity_info = self.storage.capacity_info();
+    let total_capacity = capacity_info.total();
+    let storage_end_epoch = capacity_info.capacity_end_epoch();
+    let current_epoch = system::epoch(system);
+
+    // Withdraw all available funds from stash for payment.
+    let mut payment = self.coin_stash.withdraw_wal_for_storage(available_wal, ctx);
+
+    // Phase 1: Handle dormant mode compensation if needed.
+    let compensation_epochs = if (current_epoch > storage_end_epoch) {
+        // Dormant mode: BlobManager has expired, need to compensate for past epochs.
+        // Check if still within grace period.
+        let grace_end_epoch = storage_end_epoch + self.grace_period_epochs;
+        assert!(current_epoch <= grace_end_epoch, EBeyondGracePeriod);
+
+        // Calculate compensation epochs needed (how many epochs we're past expiry).
+        let past_epochs = current_epoch - storage_end_epoch;
+
+        // Buy compensation storage for the past epochs (dormant period).
+        // We purchase this to pay for the dormant period but don't apply it.
+        let compensation_storage = system::reserve_space_for_epochs(
+            system,
+            total_capacity,
+            storage_end_epoch,
+            current_epoch,
+            &mut payment,
+            ctx,
+        );
+
+        // Destroy the compensation storage - we only needed to pay for it.
+        storage_resource::destroy(compensation_storage);
+
+        past_epochs
+    } else {
+        0
+    };
+
+    // Phase 2: Buy the actual extension storage.
+    // If we paid compensation (dormant mode), extend from current_epoch forward.
+    // Otherwise extend from the existing storage_end_epoch.
+    let extension_start_epoch = if (compensation_epochs > 0) {
+        // In dormant mode, start extension from current epoch.
+        current_epoch
+    } else {
+        // Not dormant, extend from current storage end.
+        storage_end_epoch
+    };
+    let new_end_epoch = extension_start_epoch + effective_extension;
+
+    // Purchase the extension storage.
+    let extension_storage = system::reserve_space_for_epochs(
+        system,
+        total_capacity,
+        extension_start_epoch,
+        new_end_epoch,
+        &mut payment,
+        ctx,
+    );
+
+    // Return any unused funds to the stash.
+    if (payment.value() > 0) {
+        self.coin_stash.deposit_wal(payment);
+    } else {
+        coin::destroy_zero(payment);
+    };
+
+    // Apply the extension to the BlobManager's storage.
+    self.storage.extend_storage(extension_storage);
+
+    // Tip the transaction sender for helping execute the extension.
```

Comment:
Should we always tip or only tip the public extension, and not the manager extension?